### PR TITLE
fix(qe): Prevent descriptions from mobc leaking into the presentation of metrics

### PR DESCRIPTION
**Pull request by miguelff**
This is the first of set of a series of patches to improve metrics correctness and tech-debt.

I recommend seeing the modified files entirely. The code has been clarified with comments.

In this patch, the fix is to rename the metrics descriptions properly instead of leaking the descriptions coming from mobc. 

I took the descriptions from the [docs](https://www.prisma.io/docs/concepts/components/prisma-client/metrics) and changed them for clarity. I'd love a review from @ruheni in here. You can find the descriptions of each metric in the updated smoke tests, and if you agree, maybe you can tweak the docs to reflect the new descriptions.

### CODE REVIEW

File: query-engine/black-box-tests/tests/metrics/smoke_tests.rs [hunk 1]
```diff
@@ -55,22 +55,54 @@ mod smoke_tests {
                 .text()
                 .await
                 .unwrap();
+            
+            // I would have loved to use insta in here and check the snapshot but the order of the metrics is not guaranteed
```
Comment (janpio) [inline query-engine/black-box-tests/tests/metrics/smoke_tests.rs:59]: We might want to guarantee the order of metrics in the future maybe then?

File: query-engine/black-box-tests/tests/metrics/smoke_tests.rs [hunk 2]
```diff
@@ -55,22 +55,54 @@ mod smoke_tests {
...
+            assert_eq!(metrics.matches("# TYPE prisma_pool_connections_open gauge").count(), 1);
+
+            assert_eq!(metrics.matches("# HELP prisma_client_queries_duration_histogram_ms The distribution of the time Prisma Client queries took to run end to end").count(), 1);
+            assert_eq!(metrics.matches("# TYPE prisma_client_queries_duration_histogram_ms histogram").count(), 1);
+
+            assert_eq!(metrics.matches("# HELP prisma_client_queries_wait_histogram_ms The distribution of the time all datasource queries spent waiting for a free connection").count(), 1);
+            assert_eq!(metrics.matches("# TYPE prisma_client_queries_wait_histogram_ms histogram").count(), 1);
+            
+            assert_eq!(metrics.matches("# HELP prisma_datasource_queries_duration_histogram_ms The distribution of the time datasource queries took to run").count(), 1);
+            assert_eq!(metrics.matches("# TYPE prisma_datasource_queries_duration_histogram_ms histogram").count(), 1);
```
Comment (janpio) [inline query-engine/black-box-tests/tests/metrics/smoke_tests.rs:96]:
```suggestion
            // counters
            assert_eq!(metrics.matches("# HELP prisma_client_queries_total The total number of Prisma Client queries executed").count(), 1);
            assert_eq!(metrics.matches("# TYPE prisma_client_queries_total counter").count(), 1);

            assert_eq!(metrics.matches("# HELP prisma_datasource_queries_total The total number of datasource queries executed").count(), 1);
            assert_eq!(metrics.matches("# TYPE prisma_datasource_queries_total counter").count(), 1);

            assert_eq!(metrics.matches("# HELP prisma_pool_connections_closed_total The total number of pool connections closed").count(), 1);
            assert_eq!(metrics.matches("# TYPE prisma_pool_connections_closed_total counter").count(), 1);

            assert_eq!(metrics.matches("# HELP prisma_pool_connections_opened_total The total number of pool connections opened").count(), 1);
            assert_eq!(metrics.matches("# TYPE prisma_pool_connections_opened_total counter").count(), 1);

            // gauges
            assert_eq!(metrics.matches("# HELP prisma_client_queries_active The number of currently active Prisma Client queries").count(), 1);
            assert_eq!(metrics.matches("# TYPE prisma_client_queries_active gauge").count(), 1);

            assert_eq!(metrics.matches("# HELP prisma_client_queries_wait The number of datasource queries currently waiting for an free connection").count(), 1);
            assert_eq!(metrics.matches("# TYPE prisma_client_queries_wait gauge").count(), 1);

            assert_eq!(metrics.matches("# HELP prisma_pool_connections_busy The number of pool connections currently executing datasource queries").count(), 1);
            assert_eq!(metrics.matches("# TYPE prisma_pool_connections_busy gauge").count(), 1);

            assert_eq!(metrics.matches("# HELP prisma_pool_connections_idle The number of pool connections that are not busy running a query").count(), 1);
            assert_eq!(metrics.matches("# TYPE prisma_pool_connections_idle gauge").count(), 1);

            assert_eq!(metrics.matches("# HELP prisma_pool_connections_open The number of pool connections currently open").count(), 1);
            assert_eq!(metrics.matches("# TYPE prisma_pool_connections_open gauge").count(), 1);

            // histograms
            assert_eq!(metrics.matches("# HELP prisma_client_queries_duration_histogram_ms The distribution of the time Prisma Client queries took to run end to end").count(), 1);
            assert_eq!(metrics.matches("# TYPE prisma_client_queries_duration_histogram_ms histogram").count(), 1);

            assert_eq!(metrics.matches("# HELP prisma_client_queries_wait_histogram_ms The distribution of the time all datasource queries spent waiting for a free connection").count(), 1);
            assert_eq!(metrics.matches("# TYPE prisma_client_queries_wait_histogram_ms histogram").count(), 1);
            
            assert_eq!(metrics.matches("# HELP prisma_datasource_queries_duration_histogram_ms The distribution of the time datasource queries took to run").count(), 1);
            assert_eq!(metrics.matches("# TYPE prisma_datasource_queries_duration_histogram_ms histogram").count(), 1);
```

File: query-engine/black-box-tests/tests/metrics/smoke_tests.rs [hunk 3]
```diff
@@ -55,22 +55,54 @@ mod smoke_tests {
...
+            
+            // Check that exist as many metrics as being accepted
+            let accepted_metric_count = query_engine_metrics::ACCEPT_LIST.len();
+            let displayed_metric_count = metrics.matches("# TYPE").count();
+            let non_prisma_metric_count = displayed_metric_count - metrics.matches("# TYPE prisma").count();
+
```
Comment (janpio) [inline query-engine/black-box-tests/tests/metrics/smoke_tests.rs:102]: OT: Maybe another assertion that we have 4 counters, 4 gauges and 3 histograms? Or is that repetitive?

File: query-engine/metrics/src/lib.rs [hunk 4]
```diff
@@ -34,165 +34,138 @@ use once_cell::sync::Lazy;
...
 
-// Describe all metric here so that every time for create
-// a new metric registry for a Query Instance the descriptions
-// will be in place
-pub fn describe_metrics() {
-    // counters
+/// Describe all first-party metrics that we record in prisma-engines. Metrics recorded by third-parties
+/// --like mobc-- are described by such third parties, but ignored, and replaced by the descriptions in the
+/// METRICS_RENAMES map.
+fn initialize_metrics_descriptions() {
```
Comment (janpio) [inline query-engine/metrics/src/lib.rs:119]: Nit: Maybe make more explicit that those are the "first party" metrics descriptions?

File: query-engine/metrics/src/lib.rs [hunk 5]
```diff
@@ -34,165 +34,138 @@ use once_cell::sync::Lazy;
...
-    describe_counter!(
-        PRISMA_POOL_CONNECTIONS_CLOSED_TOTAL,
-        "Total number of Pool Connections closed"
+    describe_histogram!(
+        PRISMA_CLIENT_QUERIES_DURATION_HISTOGRAM_MS,
+        "The distribution of the time Prisma Client queries took to run end to end"
+    );
+    describe_histogram!(
+        PRISMA_DATASOURCE_QUERIES_DURATION_HISTOGRAM_MS,
+        "The distribution of the time datasource queries took to run"
```
Comment (janpio) [inline query-engine/metrics/src/lib.rs:138]: Semi OT: Why are these (this one and also `PRISMA_DATASOURCE_QUERIES_TOTAL) done by us, and not mobc? Wouldn't it make sense for mobc to take care of everything `datasource` as well?

**Review by janpio — APPROVED** [review]
❤️

File: query-engine/black-box-tests/tests/metrics/smoke_tests.rs (hunk 1 above)
Comment (miguelff) [reply query-engine/black-box-tests/tests/metrics/smoke_tests.rs:59]: I thought about it, but given that's not a requirement that is set on the exposed format. I would say that sorting wouldn't add any value for consumers of metrics, and is an additional operation to make at runtime, that would (a priori) only benefit testability.

But also given that the values of the metrics will vary from run to run, (in particular histograms recording times) we would need further processing on the response to match it with a snapshot. To the least, we would need setting numeric values to a placeholder before comparing.

We could have another approach to testing, like unmarshaling into a struct and making assertions on it, but in my opinion, doing so won't add better coverage than what we have now.

**Comment by Jolg42** [comment]
Note: the test in prisma/prisma (obviously) needed to be updated, see https://github.com/prisma/prisma/pull/21096

**Pull request merged**